ext/serializer.c

```diff
@@ -1159,40 +1193,41 @@ static void _serialize_meta(zval *el, ddtrace_span_data *span, zend_string *serv
                     continue;
                 }
 
-                dd_serialize_array_meta_recursively(Z_ARRVAL(meta_zv), str_key, orig_val);
+                dd_serialize_array_meta_recursively(rust_span, str_key, orig_val);
             }
         }
         ZEND_HASH_FOREACH_END();
     }
-    meta = &meta_zv;
 
-    zval *existing_env, new_env;
-    if ((existing_env = zend_hash_str_find(Z_ARRVAL_P(meta), ZEND_STRL("env")))) {
+    zval new_env;
+    ddog_CharSlice existing_env = ddog_get_span_meta_(rust_span, dd_str_to_CharSlice("env"));
+    if (existing_env.len > 0) {
         LOG(DEPRECATED, "Using \"env\" in meta is deprecated. Instead specify the env property directly on the span.");
     } else {
         ddtrace_convert_to_string(&new_env, &span->property_env);
         if (Z_STRLEN(new_env)) {
-            existing_env = zend_hash_str_add_new(Z_ARRVAL_P(meta), ZEND_STRL("env"), &new_env);
+            ddog_add_CharSlice_span_meta_(rust_span, dd_str_to_CharSlice("env"), Z_STR_P(&new_env));
         } else {
             zval_ptr_dtor(&new_env);
-            ZVAL_EMPTY_STRING(&new_env);
-            existing_env = &new_env;
+            existing_env.ptr = "";
         }
     }
+
     if (!span->parent) {
         if (DDTRACE_G(last_flushed_root_env_name)) {
             zend_string_release(DDTRACE_G(last_flushed_root_env_name));
         }
-        DDTRACE_G(last_flushed_root_env_name) = zend_string_copy(Z_STR_P(existing_env));
+        DDTRACE_G(last_flushed_root_env_name) = dd_CharSlice_to_zend_string(existing_env);
```

Comment:
We need to refcount this. This must remain a zend_string - do the conversion to charslice where it's used.